### PR TITLE
Cleaned up: the potentially misleading comment under Event struct

### LINKED ISSUE #94602: core v1 Event API needs improvement?
**What happened**:

```yaml
apiVersion: v1
kind: Event
```

When creating the above resource (for testing), I'm getting this:

```
error: error validating "event-v1.yaml": error validating data: [ValidationError(Event): missing required field "metadata" in io.k8s.api.core.v1.Event, ValidationError(Event): missing required field "involvedObject" in io.k8s.api.core.v1.Event]; if you choose to ignore these errors, turn validation off with --validate=false
```

After checking the source code (https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L4569-L4576):

```go
type Event struct {
	metav1.TypeMeta
	// +optional
	metav1.ObjectMeta

	// Required. The object that this event is about. Mapped to events.Event.regarding
	// +optional
	InvolvedObject ObjectReference
         ...
```

The declaration says metadata is optional, which is wrong.  So the API spec generated is wrong.

The declaration of `InvolvedObject` is contradictary. It is both "Required" and "+optional".

**What you expected to happen**:

Be consistent about field definitions.

**How to reproduce it (as minimally and precisely as possible)**:

```
kubectl create -f config.yaml
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.19

### PR DESCRIPTION
**Pull request by yashvardhan-kukreja**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Cleans up and fixes all the "+optional" comments and JSON properties over the required fields of the `type Event` with respect to `core/v1`, `events.k8s.io/v1` and `v1beta1`.

**Which issue(s) this PR fixes**:

Fixes #94602 

**Special notes for your reviewer**:
As of now, I have removed the "+optional" above InvolvedObject but further we can decide for which other resources we need to such kind of cleanup again. One of the examples, is under the Pod struct where above "metadata" it says "+Optional" despite metadata being a necessary input.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -4572,7 +4572,6 @@ type Event struct {
 	metav1.ObjectMeta
 
 	// Required. The object that this event is about. Mapped to events.Event.regarding
-	// +optional
```
Comment (wojtek-t) [inline pkg/apis/core/types.go:4575]: Just to check - is that the only field that effecitvely isn't optional?

Can you please double check if fields like FirstTimestamp isn't effectively required too?
[You can do that by trying to create an Event with just metadata and InvolvedObject being set.]

Comment (yashvardhan-kukreja) [reply pkg/apis/core/types.go:4575]: Sure I ll do so.

Comment (liggitt) [reply pkg/apis/core/types.go:4575]: optional comments on the internal API types don't actually drive any behavior… is there a corresponding optional comment on the core/v1 or events/v1 external type we need to remove?

Comment (yashvardhan-kukreja) [reply pkg/apis/core/types.go:4575]: @liggitt , under core/v1, it seems pretty much sorted but in events/v1, `ObjectMeta` (metadata) was specified as "+Optional". Hence, I edited it. Apart from that, I correspondingly edited in the internal API as well.

@wojtek-t I checked and confirmed. Under event, only metadata and involvedObject keys are necessary to drive out the successful creation of the `Event` object. So, rest seems fine.

File: staging/src/k8s.io/api/events/v1/types.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ import (
 // Event is a report of an event somewhere in the cluster. It generally denotes some state change in the system.
 type Event struct {
 	metav1.TypeMeta `json:",inline"`
-	// +optional
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/api/events/v1/types.go:30]: Given you touch this file, if you create objects with the new API, more fields are actually required
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/events.go#L41

xref https://github.com/kubernetes/kubernetes/issues/94748

Can you please fix that in the same PR too?

Comment (wojtek-t) [reply staging/src/k8s.io/api/events/v1/types.go:30]: Also, please fix the v1beta1 api group.

Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:30]: @wojtek-t regarding the v1beta1.. I made the changes locally..
But regarding the events `events.k8s.io` one, I have a doubt. So, by fixing, do you mean that the strict validations should not be necessary even in case of `events.k8s.io/Event` and hence, I should edit those validation checks?
OR
Do you mean that I should just add "Required" comments in `staging/src/k8s.io/api/events/v1/types.go` under `type Event` so as to convey that the corresponding fields are actually required by the Validator function?

Comment (wojtek-t) [reply staging/src/k8s.io/api/events/v1/types.go:30]: > Do you mean that I should just add "Required" comments under type Event (inside staging/src/k8s.io/api/events/v1/types.go) so as to convey that the corresponding fields are actually required by the Validator function?

Yes - this one.

Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:30]: Alright. On it!

File: staging/src/k8s.io/api/events/v1/types.go [hunk 3]
```diff
@@ -39,28 +40,28 @@ type Event struct {
 
 	// reportingController is the name of the controller that emitted this Event, e.g. `kubernetes.io/kubelet`.
 	// This field cannot be empty for new Events.
-	// +optional
+	// required as per the strict validation checks being performed over Event resource corrersponding to events.k8s.io/v1 or newer.
```
Comment (wojtek-t) [inline staging/src/k8s.io/api/events/v1/types.go:43]: It's actually the other way around - it's not required because we added validation - we added validation because it is required (the comments were just incorrect).

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go [hunk 4]
```diff
@@ -78,8 +78,8 @@ const (
 // Event captures all the information that can be included in an API audit log.
 type Event struct {
 	metav1.TypeMeta `json:",inline"`
-	// ObjectMeta is included for interoperability with API infrastructure.
-	// +optional
+
+	// Required. The name field under metadata is necesssary.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:82]: @liggitt - should these also be changed?

**Review by wojtek-t — COMMENTED** [review]
Please also squash commits.

[And then you need to regenerate autogenerated files (please put those in a separate commit though).]

File: staging/src/k8s.io/api/events/v1/types.go (hunk 3 above)
Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:43]: ohh my bad .. will update them

Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:43]: I believe I would just add "Required". That should be enough, i suppose

**Comment by yashvardhan-kukreja** [comment]
> Please also squash commits.
> 
> [And then you need to regenerate autogenerated files (please put those in a separate commit though).]

noob question. What do you mean by regenerating the autogenerated files? How to do it? (No need to explain it if there's any existing documentation on this somewhere in the repo)

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go (hunk 4 above)
Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:82]: > // ObjectMeta is included for interoperability with API infrastructure.

Oh my bad. Removed this comment too by mistake

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go [hunk 5]
```diff
@@ -71,8 +71,8 @@ const (
 // Event captures all the information that can be included in an API audit log.
 type Event struct {
 	metav1.TypeMeta `json:",inline"`
-	// ObjectMeta is included for interoperability with API infrastructure.
-	// +optional
+
+	// Required. The name field under metadata is necesssary.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go:75]: Should we have it? It's unrelated to this change. Let's remove.

**Comment by wojtek-t** [comment]
> noob question. What do you mean by regenerating the autogenerated files? How to do it? (No need to explain it if there's any existing documentation on this somewhere in the repo)

ohh sorry
`make update`
is what should do the job for you

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go (hunk 5 above)
Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go:75]: Just to confirm, you mean to remove this comment for `..../v1beta1/types.go` and NOT for `...../v1alpha1/types.go`
(asking this because previously you mentioned to work on the v1beta1 thing )

> Also, please fix the v1beta1 api group.

File: staging/src/k8s.io/api/events/v1/types.go (hunk 2 above)
Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:30]: > > noob question. What do you mean by regenerating the autogenerated files? How to do it? (No need to explain it if there's any existing documentation on this somewhere in the repo)
> 
> ohh sorry
> `make update`
> is what should do the job for you

Faced an error with this previously and spend quite a lot of time on resolving it but couldn't work it out.
(Go version: 1.15.2)
(`uname -a`:   `Linux luci4 5.4.0-47-generic #51~18.04.1-Ubuntu SMP Sat Sep 5 14:35:50 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`)

Sorry for the huge incoming log but **the error is in the end (seemingly)**

```
► make update
# unicode/utf8
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# math/bits
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/race
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/goversion
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/nettrace
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/internal/sys
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# sync/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# internal/cpu
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# unicode
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/internal/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# runtime/cgo
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
+++ [0925 00:25:09] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
# runtime/internal/sys
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# unicode/utf8
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# math/bits
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# unicode
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/race
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/goversion
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/nettrace
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/internal/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# unicode/utf16
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/cpu
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# encoding
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# k8s.io/kubernetes/vendor/golang.org/x/mod/semver
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# k8s.io/kubernetes/vendor/golang.org/x/xerrors/internal
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# sync/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# k8s.io/kubernetes/vendor/github.com/go-logr/logr
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/cgo
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
!!! [0925 00:25:09] Call tree:
!!! [0925 00:25:09]  1: /home/yash1300/Projects/GoProjects/github.com/yashvardhan-kukreja/kubernetes/hack/lib/golang.sh:719 kube::golang::build_some_binaries(...)
!!! [0925 00:25:09]  2: /home/yash1300/Projects/GoProjects/github.com/yashvardhan-kukreja/kubernetes/hack/lib/golang.sh:863 kube::golang::build_binaries_for_platform(...)
!!! [0925 00:25:09]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0925 00:25:09] Call tree:
!!! [0925 00:25:09]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0925 00:25:09] Call tree:
!!! [0925 00:25:09]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
Makefile.generated_files:195: recipe for target '_output/bin/prerelease-lifecycle-gen' failed
make[1]: *** [_output/bin/prerelease-lifecycle-gen] Error 1
Makefile:546: recipe for target 'generated_files' failed
make: *** [generated_files] Error 2
```

**Comment by yashvardhan-kukreja** [comment]
> > noob question. What do you mean by regenerating the autogenerated files? How to do it? (No need to explain it if there's any existing documentation on this somewhere in the repo)
> 
> ohh sorry
> `make update`
> is what should do the job for you

Faced an error with this previously and spend quite a lot of time on resolving it but couldn't work it out.
(Go version: 1.15.2)
(`uname -a`:   `Linux luci4 5.4.0-47-generic #51~18.04.1-Ubuntu SMP Sat Sep 5 14:35:50 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`)

Sorry for the huge incoming log but **the error is in the end (seemingly)**

```
► make update
# unicode/utf8
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# math/bits
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/race
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/goversion
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/nettrace
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/internal/sys
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# sync/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# internal/cpu
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# unicode
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/internal/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# runtime/cgo
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
+++ [0925 00:25:09] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
# runtime/internal/sys
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# unicode/utf8
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# math/bits
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# unicode
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/race
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/goversion
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/nettrace
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/internal/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# unicode/utf16
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# internal/cpu
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# encoding
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# k8s.io/kubernetes/vendor/golang.org/x/mod/semver
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# k8s.io/kubernetes/vendor/golang.org/x/xerrors/internal
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# sync/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# k8s.io/kubernetes/vendor/github.com/go-logr/logr
compile: version "go1.14.6" does not match go tool version "go1.15.2"
# runtime/cgo
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
!!! [0925 00:25:09] Call tree:
!!! [0925 00:25:09]  1: /home/yash1300/Projects/GoProjects/github.com/yashvardhan-kukreja/kubernetes/hack/lib/golang.sh:719 kube::golang::build_some_binaries(...)
!!! [0925 00:25:09]  2: /home/yash1300/Projects/GoProjects/github.com/yashvardhan-kukreja/kubernetes/hack/lib/golang.sh:863 kube::golang::build_binaries_for_platform(...)
!!! [0925 00:25:09]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0925 00:25:09] Call tree:
!!! [0925 00:25:09]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0925 00:25:09] Call tree:
!!! [0925 00:25:09]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
Makefile.generated_files:195: recipe for target '_output/bin/prerelease-lifecycle-gen' failed
make[1]: *** [_output/bin/prerelease-lifecycle-gen] Error 1
Makefile:546: recipe for target 'generated_files' failed
make: *** [generated_files] Error 2
```

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go (hunk 5 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go:75]: the audit event is completely unrelated to the API Event. Revert this file.

For v1beta1, I meant this one:
 staging/src/k8s.io/api/events/v1beta1/types.go

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go [hunk 6]
```diff
@@ -78,8 +78,8 @@ const (
 // Event captures all the information that can be included in an API audit log.
 type Event struct {
 	metav1.TypeMeta `json:",inline"`
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/types.go:81]: Please revert this file too - it's the same as comment below.

**Comment by wojtek-t** [comment]
> Sorry for the huge incoming log but the error is in the end (seemingly)

The error is actually on the top:
```
compile: version "go1.14.6" does not match go tool version "go1.15.2"
```

Seems like some misconfiguration on your side.  Can you even build k8s locally?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go (hunk 5 above)
Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go:75]: oh alright.. got it

**Comment by yashvardhan-kukreja** [comment]
> > Sorry for the huge incoming log but the error is in the end (seemingly)
> 
> The error is actually on the top:
> 
> ```
> compile: version "go1.14.6" does not match go tool version "go1.15.2"
> ```
> 
> Seems like some misconfiguration on your side. Can you even build k8s locally?

Yeah this was some crazy misconfiguration xD, but now it's sorted.

**Anyway, I have applied the changes you suggested. 
If they're fine, then please give me a heads up, I'll squash my commits**

**Comment by wojtek-t** [comment]
We're usually not squashing autogenerated code. So that sounds fine.

@liggitt - can you please review the API?

File: staging/src/k8s.io/api/events/v1/types.go [hunk 7]
```diff
@@ -27,7 +27,8 @@ import (
 // Event is a report of an event somewhere in the cluster. It generally denotes some state change in the system.
 type Event struct {
 	metav1.TypeMeta `json:",inline"`
-	// +optional
+
+	// required.
```
Comment (liggitt) [inline staging/src/k8s.io/api/events/v1/types.go:31]: Just adding "required" as a comment does not actually make it appear as required in the openapi schema. You actually have to remove the `+optional` comment and `omitempty` tag to do that.

Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:31]: Oh alright. Thanks for the heads up. Fixing this

File: staging/src/k8s.io/api/events/v1/types.go [hunk 8]
```diff
@@ -39,29 +40,29 @@ type Event struct {
...
-	Reason string `json:"reason,omitempty" protobuf:"bytes,7,name=reason"`
+	// required.
+	Reason string `json:"reason" protobuf:"bytes,7,name=reason"`
 
 	// regarding contains the object this Event is about. In most cases it's an Object reporting controller
 	// implements, e.g. ReplicaSetController implements ReplicaSets and this event is emitted because
 	// it acts on some changes in a ReplicaSet object.
-	// +optional
-	Regarding corev1.ObjectReference `json:"regarding,omitempty" protobuf:"bytes,8,opt,name=regarding"`
+	// required.
```
Comment (liggitt) [inline staging/src/k8s.io/api/events/v1/types.go:64]: I don't see this being required by validation, revert the changes to this field?

The following event can be created successfully:
```yaml
kind: Event
apiVersion: events.k8s.io/v1
metadata:
  generateName: foo
  namespace: kube-system
eventTime: "2017-08-09T15:04:05.000000Z"
reportingInstance: foo
reportingController: foo
action: Foo
reason: Foo
type: Warning
```

**Review by liggitt — COMMENTED** [review]
On second thought, inclusion in the `required` list is used by some generated clients (in other languages) to produce clients that do not handle absent fields well. Even though new events.k8s.io/v1 Events require these fields to be set, events created via core/v1 do not, so a client reading via events.k8s.io/v1 must be able to handle absent data in those fields. I think we should revert the omitempty changes with the possible exception of metadata

**Comment by yashvardhan-kukreja** [comment]
> On second thought, inclusion in the `required` list is used by some generated clients (in other languages) to produce clients that do not handle absent fields well. Even though new events.k8s.io/v1 Events require these fields to be set, events created via core/v1 do not, so a client reading via events.k8s.io/v1 must be able to handle absent data in those fields. I think we should revert the omitempty changes with the possible exception of metadata

Cool. Thanks for the suggestion. So, just to confiirm, should I revert back the deletion of "omitempty" I did in the last commit except for ObjectMeta (metadata) ?

**Comment by liggitt** [comment]
> should I revert back the deletion of "omitempty" I did in the last commit except for ObjectMeta (metadata) ?

yes, sorry for the back and forth

**Comment by yashvardhan-kukreja** [comment]
Sure. No worries. On it

File: staging/src/k8s.io/api/events/v1/types.go (hunk 8 above)
Comment (wojtek-t) [reply staging/src/k8s.io/api/events/v1/types.go:64]: Yes - this is definitely optional.

Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:64]: Got it :) 
Actually, this fields were mentioned in the issue as supposedly required ones so I seem to follow that. My bad.
Anyways I have pushed the new changes as per your suggestions. 😊

File: staging/src/k8s.io/api/events/v1/types.go [hunk 9]
```diff
@@ -76,8 +72,8 @@ type Event struct {
 
 	// type is the type of this event (Normal, Warning), new types could be added in the future.
 	// It is machine-readable.
-	// +optional
-	Type string `json:"type,omitempty" protobuf:"bytes,11,opt,name=type"`
+	// This field cannot be empty for new Events.
+	Type string `json:"type" protobuf:"bytes,11,opt,name=type"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/events/v1/types.go:76]: restore the omitempty, core/v1 events can omit this field and so a reader of events.k8s.io/v1 events could still encounter an object that omitted this field

File: pkg/apis/core/types.go [hunk 10]
```diff
@@ -4568,11 +4568,10 @@ const (
 // TODO: Decide whether to store these separately or with the object they apply to.
 type Event struct {
 	metav1.TypeMeta
-	// +optional
+
 	metav1.ObjectMeta
 
 	// Required. The object that this event is about. Mapped to events.Event.regarding
-	// +optional
```
Comment (liggitt) [inline pkg/apis/core/types.go:4575]: revert this, an event can be created with no involved object

File: staging/src/k8s.io/api/events/v1/types.go (hunk 9 above)
Comment (yashvardhan-kukreja) [reply staging/src/k8s.io/api/events/v1/types.go:76]: got it. (damn, I m making a lot of mistakes here 🙈  )

**Comment by yashvardhan-kukreja** [comment]
ok @liggitt , I have a slightly noob question.
Can u please tell difference in usage or implementation (or which is used when) of `type Event` in `pkg/apis/core/types.go` , `staging/src/k8s.io/api/core/v1/types.go` ,  and `staging/src/k8s.io/api/events/v1`.

**Comment by wojtek-t** [comment]
@yashvardhan-kukreja - please squash the commits

**Comment by yashvardhan-kukreja** [comment]
Thanks @liggitt , @wojtek-t .. Looking forward to contributing more :)

**Comment by wojtek-t** [comment]
@liggitt - I think we should cherrypick this one - WDYT?

**Comment by liggitt** [comment]
While it's an improvement, I don't think it impacts functionality enough for a backport
